### PR TITLE
Acq 1696 ryan school association update school info interstitial to use new school data inputs

### PR DESCRIPTION
**Pull request by drizco**
This refactor aims to move the logic and state out of the presentational input components into a reusable custom hook that can be called in the parent container component, and pushing that state back down through props. This makes `SchoolDataInputs` reusable throughout the application. It also makes it easier to test along with its child components since all that's needed is to render the component with different props and assert on what is rendered. The complex logic is moved to util functions and the `useSchoolInfo` hook, which can then be tested in isolation.

There also appeared to be a bug where the school zip code was not being persisted in the database when a user selected `Click to add` for a US based school that couldn't be found via the zip code search. By making `zip` a permitted field in the `PATCH /school_infos` request, we're now able to capture the zip code and school name entered when the school was not found.
Screenshot 2024-09-09 at 9 26 19 AM

If a teacher does not teach in a school setting, they were previously never getting a school_infos entry in the db, instead they continued to be associated with the default school_infos row that only holds the US country, which also prompts the user to complete their school_infos. Or if a teacher formerly taught at a school and provided that info, if they update their school info to say they no longer teach in a school setting, it would always detect their _last complete_ school_info, so their old school. Now we've added a new school_type for `noSchoolSetting`, so we can track the teacher's actual selection and not fall back to their last complete school_info in other forms that also try to collect teacher/school association.

https://github.com/user-attachments/assets/31e3f64e-1f8a-4c0f-8534-0a043bf23bcb

## Links

Jira ticket: https://codedotorg.atlassian.net/browse/ACQ-1696

## Testing story

The existing tests were very good, but I had to move significant sections of the tests once the logic moved to the custom hook. It resulted in a substantial refactor, but it also made the presentational components (`SchoolDataInputs`, `SchoolZipSearch`, `SchoolNameInput`) much easier to test since the need to fire events and check logic pathways was greatly reduced

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/schoolInfo/SchoolInfoInterstitial.jsx
Comment (drizco) [inline apps/src/schoolInfo/SchoolInfoInterstitial.jsx]: unsure about this. I didn't see a good way to store this for the user

File: dashboard/app/views/layouts/_school_info_confirmation_dialog.html.haml
Comment (drizco) [inline dashboard/app/views/layouts/_school_info_confirmation_dialog.html.haml:19]: I do wonder if checking the user's ip address is the right thing to do here. I imagine a user could be traveling, in which case we shouldn't make assumptions about the school location

File: apps/src/schoolInfo/utils/sendAnalyticsEvent.ts
Comment (TurnerRiley) [inline apps/src/schoolInfo/utils/sendAnalyticsEvent.ts]: Is this file necessary? It seems like it takes up the same amount of space for where it's used and adds an extra step

File: apps/src/templates/SchoolDataInputs.jsx
Comment (TurnerRiley) [inline apps/src/templates/SchoolDataInputs.jsx:81]: Nit: could the `onChange` just be `e => setCountry(e)` below rather than having this `handleCountryChange`?

File: apps/src/templates/SchoolDataInputs.jsx
Comment (TurnerRiley) [inline apps/src/templates/SchoolDataInputs.jsx:86]: Same question for this

File: apps/src/templates/SchoolNameInput.jsx
Comment (TurnerRiley) [inline apps/src/templates/SchoolNameInput.jsx:13]: Same question for this (if this is for readability/consistency that's a-okay with me, just wanted to check if this was adding any functionality or not!)

File: apps/test/unit/schoolInfo/hooks/useSchoolInfoTest.jsx
Comment (TurnerRiley) [inline apps/test/unit/schoolInfo/hooks/useSchoolInfoTest.jsx]: TIL about being able to directly reference a hook in react testing!

File: apps/src/schoolInfo/utils/sendAnalyticsEvent.ts
Comment (drizco) [reply apps/src/schoolInfo/utils/sendAnalyticsEvent.ts]: The nice part of moving it to its own file is it makes it much easier to mock and spy on when testing components that use it

File: apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx
Comment (TurnerRiley) [inline apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx]: Nit: I'm not sure these need to be updated since `getByText` will throw an error if the given text is not found?

File: apps/test/unit/templates/SchoolDataInputsTest.jsx
Comment (TurnerRiley) [inline apps/test/unit/templates/SchoolDataInputsTest.jsx]: Nit: could this be pulled out into a `renderDefault` function that accepts `overrideProps` so it's clearer when scanning through tests which props are being changed to update the given output?

**Review by TurnerRiley — COMMENTED** [review]
This looks great! Left a couple small questions/nits but I'll defer to Hannah on overall approval/questions/changes since she has way more experience with the school info inputs!

File: apps/src/templates/SchoolNameInput.jsx
Comment (TurnerRiley) [reply apps/src/templates/SchoolNameInput.jsx:13]: Oh is this part of being able to mock/spy it easier?

Comment (drizco) [reply apps/src/templates/SchoolNameInput.jsx:13]: It's mostly just a convention I've gotten used to. definitely could be in the onChange event, but I like seeing all the handler functions in the body of the functional component rather than the returned jsx. Mostly because I don't want any handler logic down there (but these don't have any logic anymore 😁 )

File: apps/test/unit/schoolInfo/hooks/useSchoolInfoTest.jsx
Comment (drizco) [reply apps/test/unit/schoolInfo/hooks/useSchoolInfoTest.jsx]: for sure, I had to add a hook testing package! but yeah, it's a lot easier than testing the component that uses the hook

File: apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx
Comment (drizco) [reply apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx]: ooo good question, I'll take a look!

File: apps/test/unit/templates/SchoolDataInputsTest.jsx
Comment (drizco) [reply apps/test/unit/templates/SchoolDataInputsTest.jsx]: ah, nice call! will update to use that function

File: apps/src/schoolInfo/utils/sendAnalyticsEvent.ts
Comment (hannahbergam) [reply apps/src/schoolInfo/utils/sendAnalyticsEvent.ts]: In general, my vote would be to match the pattern of our other uses of sendEvent in our codebase. I think the additional layer of abstraction living in a different file makes it harder to understand the analytics at a glance. That said I respect this move if you have strong feelings about it!

File: apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx
Comment (drizco) [reply apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx]: I _think_ this is okay. if `getByText` throws an error it should fail the test, which is what we would expect

File: apps/test/unit/templates/SchoolDataInputsTest.jsx
Comment (drizco) [reply apps/test/unit/templates/SchoolDataInputsTest.jsx]: updated!

**Review by hannahbergam — COMMENTED** [review]
Hi! I'm going to do this in phases because of the length, but I just read through your Pr description and saw the question about the 'I don't teach in a school setting'. We currently are sending this event to statsig from two different places- it would be great to reuse the string and send it in this case as well to maintain that functionality. You can search for 'DO_NOT_TEACH_AT_SCHOOL_CLICKED'.

File: apps/src/schoolInfo/utils/sendAnalyticsEvent.ts
Comment (drizco) [reply apps/src/schoolInfo/utils/sendAnalyticsEvent.ts]: fair enough! I can remove the util function and just call the `analyticsReporter.sendEvent` method directly where it's needed

**Comment by drizco** [comment]
> Hi! I'm going to do this in phases because of the length, but I just read through your Pr description and saw the question about the 'I don't teach in a school setting'. We currently are sending this event to statsig from two different places- it would be great to reuse the string and send it in this case as well to maintain that functionality. You can search for 'DO_NOT_TEACH_AT_SCHOOL_CLICKED'.

<s>ah cool! I am sending that event in `useSchoolInfo` hook, so it must be okay to skip making the PATCH request to school_info in this case. should be good to go then</s>

No longer relevant since we found a way to send and persist school_info when the teacher does not teach in a school setting

File: apps/src/schoolInfo/SchoolInfoInterstitial.jsx
Comment (drizco) [reply apps/src/schoolInfo/SchoolInfoInterstitial.jsx]: @hannahbergam confirmed all that needs to be sent is an analytics reporter event, which is done in `useSchoolInfo`

File: apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx
Comment (TurnerRiley) [reply apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx]: Yep it does! So I think if it's okay could we keep the changes to these tests what is specific to the changes made in this PR? Hannah and I have a couple PR's open that are touching several of these files so as we add more tests and such, if we are changing the structure/patterns used it will create a ton of merge conflicts that I'm worried will accidentally sweep something under the rug that we don't mean to. I think these are good future changes/cleanups that would be great to discuss but if possible I'd like to keep the scope of this PR to a minimum so it's easier to keep moving forward with other signup changes!

File: apps/package.json
Comment (hannahbergam) [inline apps/package.json:61]: @lfryemason Just a heads up that we're moving in this direction in the acq world! @drizco for context- Liam has been leading the charge toward React Testing Library, so hoping to keep him in the loop as much as possible:)

File: apps/src/componentLibrary/dropdown/simpleDropdown/SimpleDropdown.tsx
Comment (hannahbergam) [inline apps/src/componentLibrary/dropdown/simpleDropdown/SimpleDropdown.tsx]: When I've done small refactors/updates of our design system components like this, Denys (@levadadenys ) has requested that I follow a specific procedure. Would it be alright if we pulled this part out into its own PR and tagged him in it? I believe the ideal PR would include the change, an update to the corresponding changelog, and updates of any storybook components to show how your updates work!

File: dashboard/app/views/layouts/_school_info_confirmation_dialog.html.haml
Comment (bethanyaconnor) [reply dashboard/app/views/layouts/_school_info_confirmation_dialog.html.haml:19]: That's definitely a possibility, but users can still choose the US as their country manually and it would still work the same way (I believe)

File: apps/package.json
Comment (lfryemason) [reply apps/package.json:61]: Are we able to use `renderHooks` from `@testing-library/react` instead of the standalone library? [It seems like they recently added it (corrected link)](https://github.com/testing-library/react-hooks-testing-library?tab=readme-ov-file#a-note-about-react-18-support)

Comment (bethanyaconnor) [reply apps/package.json:61]: Is that the correct link @lfryemason? I'm taken to a tech spec.

Comment (lfryemason) [reply apps/package.json:61]: It absolutely is not: https://github.com/testing-library/react-hooks-testing-library?tab=readme-ov-file#a-note-about-react-18-support

File: apps/src/componentLibrary/dropdown/simpleDropdown/SimpleDropdown.tsx
Comment (drizco) [reply apps/src/componentLibrary/dropdown/simpleDropdown/SimpleDropdown.tsx]: for sure, that makes sense! I'll pull this out

File: apps/package.json
Comment (drizco) [reply apps/package.json:61]: it's not in our current version of react testing library, looks like we're on 12.1.5 

I can try updating though

Comment (drizco) [reply apps/package.json:61]: it seems like bumping to 13.x to get `renderHook` doesn't work for react versions under 18 unfortunately. are we considering a react update to 18 soon?

File: apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx
Comment (drizco) [reply apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx]: yep, that's fair! I'll spend some time tomorrow morning trying to scope this down, though it will be tricky if we're trying to get away from relying on jquery and session storage to access the form state from all over the place. I'm happy to jump on a call to discuss the changes in more detail since I absolutely did not want to step on toes or disrupt ongoing work! in fact it would be really great to get a tour of other places this component is planned to be used. I definitely should have asked for that before I got too deep on this! 😅

Comment (drizco) [reply apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx]: ah, I misunderstood your first point. I was confused about not seeing expect wrapping those `getByText`. I get it now, makes sense! I'll put this back the way I found it 😬

Comment (TurnerRiley) [reply apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx]: Thank you! No for sure, thank you for your insight and I think this would be a great thing to discuss and cleanup soon! Definitely didn't step on any toes :D

Comment (drizco) [reply apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx]: hoho, I had kind of forgotten why I even touched these tests in the first place! it was late on friday and I was getting frustrated with a json circular reference error I was getting instead of a test failure when running the first two tests together (or the whole suite). when run individually, each test passed fine.

the circular reference error was a red herring I'm still a bit confused about. but ultimately it was finding the zip code element with queryByText instead of null. the issue was `SCHOOL_COUNTRY_SESSION_KEY` was not being reset in sessionStorage in the afterEach block, which was taking precedence over the usIp prop. 

anyway, should be all good now!

Comment (TurnerRiley) [reply apps/test/unit/signUpFlow/FinishTeacherAccountTest.tsx]: Ah gotcha gotcha!

**Review by hannahbergam — COMMENTED** [review]
Hi! One small last thing to make sure looks alright before we do the final review tomorrow- I know you've been setting up a disabled submit state for the school info interstitial- can we make sure folks can still finish a new account sign up with incomplete school information? We have in our roadmap to AB test a sign up flow where school information is required, but for now we want to make sure it isn't required. I'm specifically imagining a state where a user has written a zip but not selected from the dropdown (we don't want to block a save in this case).

File: dashboard/app/models/school_info.rb
Comment (drizco) [inline dashboard/app/models/school_info.rb]: here's the new school type for non-school settings

File: dashboard/app/views/layouts/_school_info_confirmation_dialog.html.haml
Comment (drizco) [inline dashboard/app/views/layouts/_school_info_confirmation_dialog.html.haml:21]: Kind of hard to tell what's happening here.. So when an nces school with all its info is associated with a teacher, we use that zip, which is stored as a string. When a teacher can't find their school, or they don't teach in a school setting, we store the zip they provide, but the school_info.zip column is an integer instead of a string. Rather than running a migration to fix this column, I'm converting to a string and padding the start with zeros 🤷

**Comment by drizco** [comment]
> Hi! One small last thing to make sure looks alright before we do the final review tomorrow- I know you've been setting up a disabled submit state for the school info interstitial- can we make sure folks can still finish a new account sign up with incomplete school information? We have in our roadmap to AB test a sign up flow where school information is required, but for now we want to make sure it isn't required. I'm specifically imagining a state where a user has written a zip but not selected from the dropdown (we don't want to block a save in this case).

For sure! the disabled submit button only exists in the `SchoolInfoInterstitial` component, not the `SchoolDataInputs` component rendered in `_finish_sign_up.js`, so it's not affected by that logic. I've manually tested both sign up and updating school info via the interstitial. I think we should also go through all the possible flows manually during the pr review to be 💯 sure we're good to go

File: apps/src/schoolInfo/SchoolInfoInterstitial.jsx
Comment (bethanyaconnor) [inline apps/src/schoolInfo/SchoolInfoInterstitial.jsx:135]: Should we count attempts rather than assuming?

File: dashboard/app/models/school_info.rb
Comment (bethanyaconnor) [reply dashboard/app/models/school_info.rb]: tiny tiny nit: should this be before other? I feel like "other" is always at the end of these lists.

Not worth updating if there aren't other changes to make. And definitely not blocking either way

**Review by bethanyaconnor — APPROVED** [review]
Played with it locally and all looked good. I'd like Hannah to approve before merging though!

The one change I see happening here is when a user says their school isn't listed, we no longer store city and state, right? Not sure if that matters given, but it would cause a small amount of data inconsistency.

File: apps/src/schoolInfo/SchoolInfoInterstitial.jsx
Comment (drizco) [reply apps/src/schoolInfo/SchoolInfoInterstitial.jsx:135]: haha, this is one of the very few things I actually didn't change from the original implementation 😅

**Comment by drizco** [comment]
> Played with it locally and all looked good. I'd like Hannah to approve before merging though!
> 
> The one change I see happening here is when a user says their school isn't listed, we no longer store city and state, right? Not sure if that matters given, but it would cause a small amount of data inconsistency.

Yeah, I assumed that was part of the original `finish_sign_up` refactor that created the `SchoolDataInputs` component. it didn't have city or state, just zip.

File: apps/src/schoolInfo/SchoolInfoInterstitial.jsx
Comment (bethanyaconnor) [reply apps/src/schoolInfo/SchoolInfoInterstitial.jsx:135]: LOL sorry, then definitely no need to fix 😅

**Comment by bethanyaconnor** [comment]
oh true 🤦‍♀️

**Comment by drizco** [comment]
@bethanyaconnor a test in `FinishTeacherAccountTest` was failing after rebasing with staging, so I've made that update after you approved. I also refactored out some unnecessary data passed down from the haml files for making the api request, which I don't have in the account settings page. (my other school info ticket is branched off of this)

I'll see if @hannahbergam and @TurnerRiley are available monday to review again 😅

**Review by TurnerRiley — APPROVED** [review]
LGTM! Though I can't really speak to the now-no-longer-needed data for the api request, @hannahbergam has more insight on that!

**Review by hannahbergam — APPROVED** [review]
We walked through this again offline- LGTM! Happy to go with collecting the zip for unlisted schools or not pending your communication with RED.

**Comment by drizco** [comment]
> We walked through this again offline- LGTM! Happy to go with collecting the zip for unlisted schools or not pending your communication with RED.

I removed it... 🤷

**Pull request merged**
